src/Microsoft.ML.Recommender/MatrixFactorizationPredictor.cs

```diff
@@ -0,0 +1,496 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Internallearn;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.Runtime.Model;
+using Microsoft.ML.Runtime.Recommender;
+using Microsoft.ML.Runtime.Recommender.Internal;
+using Microsoft.ML.Runtime.Model.Pfa;
+using Microsoft.ML.Runtime.Model.Onnx;
+using Microsoft.ML.Runtime.CommandLine;
+using Microsoft.ML.Runtime.Data.IO;
+
+[assembly: LoadableClass(typeof(MatrixFactorizationPredictor), null, typeof(SignatureLoadModel), "Matrix Factorization Predictor Executor", MatrixFactorizationPredictor.LoaderSignature)]
+
+[assembly: LoadableClass(typeof(MatrixFactorizationPredictionTransformer), typeof(MatrixFactorizationPredictionTransformer),
+    null, typeof(SignatureLoadModel), "", MatrixFactorizationPredictionTransformer.LoaderSignature)]
+
+namespace Microsoft.ML.Runtime.Recommender
+{
+    public sealed class MatrixFactorizationPredictor : IPredictor, ICanSaveModel, ICanSaveInTextFormat, ISchemaBindableMapper
+    {
+        internal const string LoaderSignature = "MFPredictor";
+        internal const string RegistrationName = "MatrixFactorizationPredictor";
+
+        private static VersionInfo GetVersionInfo()
+        {
+            return new VersionInfo(
+                modelSignature: "FAFAMAPD",
+                verWrittenCur: 0x00010001,
+                verReadableCur: 0x00010001,
+                verWeCanReadBack: 0x00010001,
+                loaderSignature: LoaderSignature,
+                loaderAssemblyName: typeof(MatrixFactorizationPredictor).Assembly.FullName);
+        }
+
+        private readonly IHost _host;
+        // The number of rows.
+        private readonly int _m;
+        // The number of columns.
+        private readonly int _n;
+        // The internal dimension.
+        private readonly int _k;
+        // The libMF is always single precision, so we will
+        // keep it as such even in the double build, even when
+        // the ML.NET output type is noted as being double.
+
+        // Packed _m by _k matrix.
+        private readonly float[] _p;
+        // Packed _k by _n matrix.
+        private readonly float[] _q;
```

Comment:
_rightComponent, _rightSubmatrix?

Reply:
No. I like to use the notations in LIBMF paper.